### PR TITLE
Amendment: `featureRemit`

### PR DESCRIPTION
**Pull request by dangell7**
### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [x] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

### CODE REVIEW

**Comment by tequdev** [comment]
Is it expected that the transaction will succeed if none of the fields sfAmounts, sfURITokenIDs, or sfMintURIToken are specified?

From the code it looks like it will succeed, but it may be appropriate that it will be an error.

It's okay if success is expected.

**Comment by dangell7** [comment]
> Is it expected that the transaction will succeed if none of the fields sfAmounts, sfURITokenIDs, or sfMintURIToken are specified?
> 
> From the code it looks like it will succeed, but it may be appropriate that it will be an error.
> 
> It's okay if success is expected.

It will create the account if the account does not exist.

File: src/ripple/app/hook/impl/applyHook.cpp [hunk 1]
```diff
@@ -70,6 +70,17 @@ getTransactionalStakeHolders(STTx const& tx, ReadView const& rv)
...
     {
+
+        case ttREMIT: {
+            if (destAcc)
+                ADD_TSH(*destAcc, tshSTRONG);
+
+            if (tx.isFieldPresent(sfInform))
+                ADD_TSH(tx.getAccountID(sfInform), tshWEAK);
+
+            break;
```
Comment (tequdev) [inline src/ripple/app/hook/impl/applyHook.cpp]: If URIToken is included, shouldn't the issuer be set to TSH?

Comment (dangell7) [reply src/ripple/app/hook/impl/applyHook.cpp:81]: Issuer is the OTXN Account and so by default, the strong TSH.

Comment (dangell7) [reply src/ripple/app/hook/impl/applyHook.cpp]: Yes it should. Thank you

**Review by tequdev — COMMENTED** [review]
If the account specified for sfInform is a Destination or an issuing account with a Burnable URIToken issued by a third party, will the Inform account fire the Hook as Strong or Weak?

I don't know how cases where the same account is registered multiple times with ADD_TSH are processed.

**Comment by dangell7** [comment]
> If the account specified for sfInform is a Destination or an issuing account with a Burnable URIToken issued by a third party, will the Inform account fire the Hook as Strong or Weak?
> 
> I don't know how cases where the same account is registered multiple times with ADD_TSH are processed.

I'm more in favor of removing inform. It's not allowed to be the account or destination. I will fix this. 

It would fire as strong, if set to the issuer, with burnable, else weak I believe, I will double check this.

File: src/ripple/protocol/impl/InnerObjectFormats.cpp [hunk 2]
```diff
@@ -143,6 +143,21 @@ InnerObjectFormats::InnerObjectFormats()
             {sfPublicKey, soeREQUIRED},
             {sfAccount, soeOPTIONAL},
         });
+
+    add(sfAmountEntry.jsonName.c_str(),
+        sfAmountEntry.getCode(),
+        {
+            {sfAmount, soeREQUIRED},
+            {sfFlags, soeOPTIONAL},
```
Comment (tequdev) [inline src/ripple/protocol/impl/InnerObjectFormats.cpp]: What is it used for?

Comment (dangell7) [reply src/ripple/protocol/impl/InnerObjectFormats.cpp]: Informing a 3rd party

Comment (dangell7) [reply src/ripple/protocol/impl/InnerObjectFormats.cpp]: I read that wrong, inform is for 3rd party. The flag here is unused. I think we were discussing the option and never removed it. I will remove it now.

File: src/ripple/protocol/impl/TxFormats.cpp [hunk 3]
```diff
@@ -116,6 +116,21 @@ TxFormats::TxFormats()
...
+        {
+            {sfDestination, soeREQUIRED},
+            {sfAmounts, soeOPTIONAL},
+            {sfURITokenIDs, soeOPTIONAL},
+            {sfMintURIToken, soeOPTIONAL},
+            {sfInvoiceID, soeOPTIONAL},
+            {sfDestinationTag, soeOPTIONAL},
+            {sfTicketSequence, soeOPTIONAL},
+            {sfBlob, soeOPTIONAL},
+            {sfInform, soeOPTIONAL},
```
Comment (tequdev) [inline src/ripple/protocol/impl/TxFormats.cpp:130]: Do you Keep sfInform Field?

File: src/ripple/protocol/impl/InnerObjectFormats.cpp (hunk 2 above)
Comment (tequdev) [inline src/ripple/protocol/impl/InnerObjectFormats.cpp]: Do you keep sfFlags in sfAmountEntry?

File: src/ripple/app/tx/impl/Remit.cpp [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+    XRPAmount nativeRemit{0};
+
+    AccountID const dstAccID{ctx_.tx[sfDestination]};
+    auto sleDstAcc = sb.peek(keylet::account(dstAccID));
+    auto const flags = !sleDstAcc ? 0 : sleDstAcc->getFlags();
+
+    // Check if the destination has disallowed incoming
+    if (sb.rules().enabled(featureDisallowIncoming) &&
+        (flags & lsfDisallowIncomingRemit))
+        return tecNO_PERMISSION;
```
Comment (tequdev) [inline src/ripple/app/tx/impl/Remit.cpp:276]: Is lsfDisallowIncomingRemit flag finally opt-out? I'd like to know the reason?

File: src/ripple/protocol/impl/TxFormats.cpp (hunk 3 above)
Comment (dangell7) [reply src/ripple/protocol/impl/TxFormats.cpp:130]: Yes Inform will stay.

File: src/ripple/protocol/impl/InnerObjectFormats.cpp (hunk 2 above)
Comment (dangell7) [reply src/ripple/protocol/impl/InnerObjectFormats.cpp]: No I will remove this.

File: src/ripple/app/tx/impl/Remit.cpp (hunk 4 above)
Comment (dangell7) [reply src/ripple/app/tx/impl/Remit.cpp:276]: Yes it is opt out. The goal of ttRemit is so that the receiving user doesn't have to perform any action.

File: src/ripple/app/tx/impl/Remit.cpp [hunk 5]
```diff
@@ -0,0 +1,545 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp:4]: these copyrights are wrong and need to be updated in a number of places, but this PR isn't that place

File: src/ripple/app/tx/impl/Remit.cpp [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+#include <ripple/basics/Log.h>
+#include <ripple/ledger/View.h>
+#include <ripple/protocol/Feature.h>
+#include <ripple/protocol/Indexes.h>
+namespace ripple {
+
+TxConsequences
+Remit::makeTxConsequences(PreflightContext const& ctx)
+{
+    return TxConsequences{ctx.tx, TxConsequences::normal};
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp]: this is probably wrong, see the payment transactor tx consequences https://github.com/Xahau/xahaud/blob/caa8b382d8dfbf3bfd14e183e2e9cf10e68ffc0a/src/ripple/app/tx/impl/Payment.cpp#L31-L44

File: src/ripple/app/tx/impl/Remit.cpp [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+    AccountID const dstID = ctx.tx.getAccountID(sfDestination);
+    AccountID const srcID = ctx.tx.getAccountID(sfAccount);
+
+    if (dstID == srcID)
+    {
+        JLOG(ctx.j.warn()) << "Malformed transaction: Remit to self.";
+        return temREDUNDANT;
+    }
+
+    if (ctx.tx.getFieldVL(sfBlob).size() > (128 * 1024))
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp]: sfBlob is optional but that's not checked here, should be if (ctx.tx.isFieldPresent(sfBlob) &&...

File: src/ripple/app/tx/impl/Remit.cpp [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+                           << ctx.tx.getTransactionID();
+        return temMALFORMED;
+    }
+
+    // sanity check amounts
+    if (ctx.tx.isFieldPresent(sfAmounts))
+    {
+        if (ctx.tx.getFieldArray(sfAmounts).size() > 32)
+        {
+            JLOG(ctx.j.warn()) << "Malformed: AmountEntrys Exceed Limit `32`.";
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp]: AmountEntry count exceeds

File: src/ripple/app/tx/impl/Remit.cpp [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+        }
+    }
+
+    // sanity check minturitoken
+    if (ctx.tx.isFieldPresent(sfMintURIToken))
+    {
+        STObject const& mint = const_cast<ripple::STTx&>(ctx.tx)
+                                   .getField(sfMintURIToken)
+                                   .downcast<STObject>();
+        // RH TODO: iterate mint fields detect any that shouldnt be there
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp]: should be done like here: https://github.com/Xahau/xahaud/blob/caa8b382d8dfbf3bfd14e183e2e9cf10e68ffc0a/src/ripple/app/tx/impl/SetHook.cpp#L716-L732

File: src/ripple/app/tx/impl/Remit.cpp [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+
+        Blob const uri = mint.getFieldVL(sfURI);
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp:183]: not sure how well inner object template system works, we should probably guard this with a isFieldPresent to be safe

File: src/ripple/app/tx/impl/Remit.cpp [hunk 11]
```diff
@@ -0,0 +1,545 @@
...
+        {
+            if (mint.getFieldU32(sfFlags) & tfURITokenMintMask)
+                return temINVALID_FLAG;
+        }
+    }
+
+    // sanity check uritokenids
+    if (ctx.tx.isFieldPresent(sfURITokenIDs))
+    {
+        if (ctx.tx.getFieldV256(sfURITokenIDs).size() > 32)
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp]: Not sure if vec256 size 0 is possible but we should probably defensively check here

File: src/ripple/app/tx/impl/Remit.cpp [hunk 12]
```diff
@@ -0,0 +1,545 @@
...
+    auto sleDstAcc = sb.peek(keylet::account(dstAccID));
+    auto const flags = !sleDstAcc ? 0 : sleDstAcc->getFlags();
+
+    // Check if the destination has disallowed incoming
+    if (sb.rules().enabled(featureDisallowIncoming) &&
+        (flags & lsfDisallowIncomingRemit))
+        return tecNO_PERMISSION;
+
+    // Check if the destination account requires deposit authorization.
+    bool const depositAuth{view().rules().enabled(featureDepositAuth)};
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp]: should be sb.rules(). one view only in use at a time

File: src/ripple/app/tx/impl/Remit.cpp [hunk 13]
```diff
@@ -0,0 +1,545 @@
...
+    if (depositAuth && sleDstAcc && (sleDstAcc->getFlags() & lsfDepositAuth))
+    {
+        if (!view().exists(keylet::depositPreauth(dstAccID, srcAccID)))
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp]: sb.exists

File: src/ripple/app/tx/impl/Remit.cpp [hunk 14]
```diff
@@ -0,0 +1,545 @@
...
+        sleMint->setFieldVL(sfURI, mintURI);
+
+        if (mint.isFieldPresent(sfDigest))
+            sleMint->setFieldH256(sfDigest, mint.getFieldH256(sfDigest));
+
+        sleMint->setFieldU32(
+            sfFlags,
+            mint.isFieldPresent(sfFlags) ? mint.getFieldU32(sfFlags) : 0);
+
+        auto const page = view().dirInsert(
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp]: should be sb.dirInsert?

File: src/ripple/app/tx/impl/Remit.cpp [hunk 15]
```diff
@@ -0,0 +1,545 @@
...
+                     canXfer != tesSUCCESS)
+                return canXfer;
+
+            // compute the amount the source will need to send
+            // in remit the sender pays all transfer fees, so that
+            // the destination can always be assured they got the exact amount
+            // specified. therefore we need to compute the amount + transfer fee
+            auto const srcAmt =
+                issuerAccID != srcAccID && issuerAccID != dstAccID
+                ? multiply(amount, transferRate(sb, issuerAccID))
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp:523]: I don't think this transfer rate is tested anywhere? given history of getting this wrong please add test cases for 0 xfer rate and non-zero as well

**Review by RichardAH — CHANGES_REQUESTED** [review]

File: src/ripple/app/tx/impl/Remit.cpp (hunk 5 above)
Comment (dangell7) [reply src/ripple/app/tx/impl/Remit.cpp:4]: https://github.com/Xahau/xahaud/issues/282

File: src/ripple/app/tx/impl/Remit.cpp (hunk 15 above)
Comment (dangell7) [reply src/ripple/app/tx/impl/Remit.cpp:523]: I tested it the same way we did in URIToken after the fix. https://github.com/Xahau/xahaud/blob/4e8c0b4388eaac006e71dd7dd9bca4a0c73e5d5c/src/test/app/Remit_test.cpp#L2017

Will add test for 0 and non 0.

File: src/ripple/app/tx/impl/Remit.cpp (hunk 10 above)
Comment (dangell7) [reply src/ripple/app/tx/impl/Remit.cpp:183]: The submit throws if you don't include the URI field. Also throws if you include a field not present in the template. Either way I still added the defensive checks.

File: src/ripple/app/hook/impl/applyHook.cpp [hunk 16]
```diff
@@ -70,6 +70,39 @@ getTransactionalStakeHolders(STTx const& tx, ReadView const& rv)
 
     switch (tt)
     {
+        case ttREMIT: {
+            if (destAcc)
+                ADD_TSH(*destAcc, tshSTRONG);
+
+            if (tx.isFieldPresent(sfInform))
```
Comment (RichardAH) [inline src/ripple/app/hook/impl/applyHook.cpp:77]: We should check if the sfInform is the same as destination or origin

We should also check if the sfInform exists (in doApply) and it should be an error if it does not.

File: src/ripple/app/hook/impl/applyHook.cpp [hunk 17]
```diff
@@ -70,6 +70,39 @@ getTransactionalStakeHolders(STTx const& tx, ReadView const& rv)
...
+                        continue;
+
+                    auto const ut = rv.read(id);
+                    if (!ut ||
+                        ut->getFieldU16(sfLedgerEntryType) != ltURI_TOKEN)
+                        continue;
+
+                    auto const owner = ut->getAccountID(sfOwner);
+                    auto const issuer = ut->getAccountID(sfIssuer);
+                    if (issuer != owner)
```
Comment (RichardAH) [inline src/ripple/app/hook/impl/applyHook.cpp]: the issuer might also be the same as the destination

File: src/ripple/app/tx/impl/Remit.cpp [hunk 18]
```diff
@@ -0,0 +1,560 @@
...
+    auto const flags = !sleDstAcc ? 0 : sleDstAcc->getFlags();
+
+    // Check if the destination has disallowed incoming
+    if (sb.rules().enabled(featureDisallowIncoming) &&
+        (flags & lsfDisallowIncomingRemit))
+        return tecNO_PERMISSION;
+
+    // Check if the destination account requires deposit authorization.
+    bool const depositAuth{sb.rules().enabled(featureDepositAuth)};
+    if (depositAuth && sleDstAcc && (sleDstAcc->getFlags() & lsfDepositAuth))
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp]: sleDstAcc->getFlags() should just be flags, it's set above

File: src/ripple/app/tx/impl/Remit.cpp [hunk 19]
```diff
@@ -0,0 +1,560 @@
...
+XRPAmount
+Remit::calculateBaseFee(ReadView const& view, STTx const& tx)
+{
+    XRPAmount extraFee{0};
+
+    if (tx.isFieldPresent(sfBlob))
+        extraFee +=
+            XRPAmount{static_cast<XRPAmount>(tx.getFieldVL(sfBlob).size())};
+
+    // RH TODO: add fees
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp]: not sure what other fees are left to add, this comment can probably go away

**Review by RichardAH — CHANGES_REQUESTED** [review]

File: src/ripple/app/tx/impl/Remit.cpp [hunk 20]
```diff
@@ -0,0 +1,592 @@
...
+            // if the target trustline doesn't exist we need to create it and
+            // pay its reserve
+            if (issuerAccID != dstAccID &&
+                !sb.exists(
+                    keylet::line(dstAccID, issuerAccID, amount.getCurrency())))
+                nativeRemit += objectReserve;
+
+            // action the transfer
+            if (TER result =
+                    accountSend(sb, srcAccID, dstAccID, amount, j, false);
```
Comment (RichardAH) [inline src/ripple/app/tx/impl/Remit.cpp]: should this be (..., true)?

**Review by RichardAH — APPROVED** [review]
